### ISSUE TITLE
Segmentation fault with sycl::buffer(InputIt, InputIt) constructor

### ISSUE THREAD

**Issue by arkantos493** [illuhad/hipSYCL#272]
```cpp
#include <vector>
#include <CL/sycl.hpp>

namespace sycl = cl::sycl;

struct A {
    sycl::buffer<float, 1> buffer;
    A(std::vector<float>& vec) : buffer(vec.begin(), vec.end()) { }    
//    A(std::vector<float>& vec) : buffer(vec.size()) {
//        auto acc = buffer.get_access<sycl::access::mode::discard_write>();
//        for (std::size_t i = 0; i < vec.size(); ++i) {
//            acc[i] = vec[i];
//        }
//    }    
};

auto make_A(const std::size_t size, const std::size_t dims) {
    std::vector<float> buf(size * dims);
    auto ptr = new A(buf);
    return ptr;
}

int main() {   
    auto data = make_A(100, 5);
//    auto data = make_A(10, 5);
    delete data;
    return 0;
}
```
Executing this code (compiled with `syclcc-clang --hipsycl-platform=cpu --std=c++17 -O3 main.cpp`) results in:
```bash
$ ./a.out                                                     
free(): corrupted unsorted chunks
[1]    16889 abort (core dumped)  ./a.out 
```

However it works if I change the size from 100 to 10 **or** if I use the size constructor for the `sycl::buffer` and manually copy all elements to it.

But to my understanding this code as given shouldn't result in a segmentation fault or am I wrong?

**Comment by illuhad** [comment]
The issue is that that particular section of the specification was reworded recently and hipSYCL still complies with an older formulation that is more open to interpretation. Basically, in hipSYCL buffers constructed using iterator pairs perform writebacks just like buffers initialized with host pointers do.

In your code, the memory described by the iterator pairs becomes invalid at the end of `make_A`. But the buffer is only destroyed at the end of `main` whey you `delete data`. When it finally gets destroyed, it then attempts to write back the buffer state to the iterator pair, and crashes because the vector is of course already gone.

You can work around that issue by explicitly disabling writeback using the `set_write_back(false)` buffer member function.

In general, I would however strongly advise against using buffer initialization with iterator pairs because this can have negative performance implications as it requires in general an additional host copy of the data, both for the initialization and potentially the write back.
This is particularly relevant when, as in your example, targeting CPU. The hipSYCL CPU backend is smart enough to realize that the host memory space is the same as the memory space of the compute device. If you initialize a buffer with a non-const pointer and target CPU, hipSYCL will therefore let compute kernels operate directly on the user-provided memory, with zero copies in between. This does not happen when using iterators.
While iterator buffer construction could maybe be optimized for some special cases, in general performance when using them will be unpredictable from the perspective of most users. Of course, we use technologies like SYCL to be faster than pure C++, so in the end we want predictable performance advantages :)

**Comment by arkantos493** [comment]
Thanks for the detailed explanation. It works as intended with `set_write_back(false)`.

I agree with you that copying isn't really optimal. But afaik I have to copy the vector because later (while the buffer still exists) I have to send the data to other nodes using MPI. This wouldn't work with the other constructors that take ownership of the underlying data. Or am I wrong?

**Comment by illuhad** [comment]
> I agree with you that copying isn't really optimal. But afaik I have to copy the vector because later (while the buffer still exists) I have to send the data to other nodes using MPI. This wouldn't work with the other constructors that take ownership of the underlying data. Or am I wrong?

If you want to send the buffer content and you cannot guarantee that the original vector still exists at the point of the MPI communication, you have to let SYCL manage the lifetime of the data, yes. However there are two major ways to achieve this without using iterators:
1. If you initialize a SYCL buffer of non-const data type with const-pointer and range, the SYCL implementation is forced to only use this pointer for initialization and copy its content to an internal storage. In this scenario, we have data initialization by the user in some external const pointer followed by a copy.
2. If you pass a SYCL buffer only a range, it will also allocate some internal storage. You can then directly initialize the content of this internal storage by getting access to it with a host accessor. In this scenerio, you can potentially eliminate the copy because the data initialization can take place directly in the internal SYCL buffer storage. Because the SYCL implementation is free to do the allocation itself, it might also perform additional optimizations such as using page-locked/pinned memory when targeting GPUs.

If you use iterators, an *additional* copy might be necessary because iterators can potentially describe non-contiguous memory regions. So, in the worst case you might end up with: User data initialization in vector -> SYCL copies memory described by iterators to contiguous memory region -> SYCL copies contiguous memory region to internal buffer. On CPU, one of those copies could be optimized away because the temporary contiguous memory region can potentially be reused as the internal buffer storage of the compute device, but I don't think hipSYCL does that currently.
In summary, I would always recommend methods 1 and 2 over iterators, with a preference on method 2.

**Comment by arkantos493** [comment]
Thanks for the detailed explanation. I guess I will use methode 2 then.
(Maybe in C++20 the iterator case can be implemented more efficient using the new `std::contiguous_iterator` concept.)

One last question (which isn't directly connected to this issue): Is it possible to tell SYCL on which GPU (in case of a multi-GPU system) the memory should be allocated?

**Comment by illuhad** [comment]
> (Maybe in C++20 the iterator case can be implemented more efficient using the new std::contiguous_iterator concept.)

Yes, although we might instead just want to use something based on ranges by that time...

> One last question (which isn't directly connected to this issue): Is it possible to tell SYCL on which GPU (in case of a multi-GPU system) the memory should be allocated?

General answer from the point of view of the SYCL spec: No, and it's not necessary. A buffer, as in the specification, is an object that manages memory on multiple devices, including host. This means that if you use a SYCL buffer inside a command group that is executing on a specific device, the buffer needs to make sure that an allocation on that device exists *at the point of kernel exection*. But since the buffer cannot know at construction on which device the data is going to be used, SYCL is usually implemented with a lazy memory allocation mechanism, where memory on the target device is only allocated when the buffer is used on that particular device. This means that a buffer will in general have to manage multiple device allocations if it's used on multiple devices. Don't think of a buffer as an abstraction layer for a single device memory allocation. It's not, it manages device allocations on as many devices as needed.

hipSYCL specific answer: In the current hipSYCL runtime, there are some substantial limitations regarding the process outlined above. The current hipSYCL buffer is hardcoded to mange an host-side data buffer and exactly one device-side data buffer, both of which are allocated at the time of buffer construction (unless you give it some already existing host buffer as pointer, as discussed above). This means that you can use `hipSetDevice()/cudaSetDevice()` before buffer construction to force allocation on a specific device. You can also tell it to allocate CUDA unified memory when targeting NVIDIA GPUs which will make the buffer available everywhere. Note that large portions of hipSYCL are currently being rewritten pretty much from scratch, which will result in a brand new runtime library. This new runtime will then also transition to lazy memory allocation and automatic management of multiple allocations.
If MPI is involved, it might be a good idea to use 1 MPI process per GPU and define which GPU to use using the environment variables `CUDA_VISIBLE_DEVICES/HIP_VISIBLE_DEVICES`.
